Add repository policy to label and close stale issues

jeffhandley opened the pull request.

jeffhandley committed:
Add repository policy to label and close stale issues

MiYanni commented on .github/policies/close-stale-issues.yml:
```diff
@@ -0,0 +1,55 @@
+id: close-stale-issues
```
I'd just add a comment with a link to the docs, since this bot system is fairly new.

```suggestion
# Resource Management policy docs: https://eng.ms/docs/more/github-inside-microsoft/policies/resource-management

id: close-stale-issues
```

marcpopMSFT commented on .github/policies/close-stale-issues.yml:
```diff
@@ -0,0 +1,55 @@
+    eventResponderTasks:
+      - description: Remove 'stale' label from issues when a comment is added
+        if:
+          - payloadType: Issue_Comment
```
Could one of us manually add back the stale label in triage to reset the clock?

jeffhandley replied:
Correct; any comment will reset the 6-year clock. _But yes_, adding the `stale` label back then bypasses the 6-year clock and restarts the 30-day clock. Or, the issue could just be manually closed of course.

marcpopMSFT commented on .github/policies/close-stale-issues.yml:
```diff
@@ -0,0 +1,55 @@
+        frequencies:
+          - weekday:
+              day: Wednesday
+              time: 16:00
```
any reason you picked Monday 12 and Wednesday 16 for these?

jeffhandley replied:
Monday so that the comment is laid down at the beginning of the week. Wednesday so that it would be 30 days (and one hour) after the comment is laid down and the label is applied. This puts both events in the middle of the work week.

MiYanni reviewed (approved):
👍

danmoseley commented:
just curious why we didn't use https://github.com/actions/stale ?

jeffhandley commented:
> just curious why we didn't use https://github.com/actions/stale ?

That's a solid question. Thus far, inertia has pulled us into using Policy Service for features it supports. I'm supportive of instead using [actions/stale](https://github.com/actions/stale) and configuring it to only apply to issues (not PRs) by using `-1` for both `days-before-pr-stale` and `days-before-pr-close`.

marcpopMSFT committed:
add a comment

Co-authored-by: Michael Yanni <MiYanni@microsoft.com>

marcpopMSFT commented:
@jeffhandley I see the comments above about stale. Since we have this PR in place and ready, is there a reason not to just keep with this? Are other teams going to switch to using actions for this?

jeffhandley commented:
There've been a couple other chats over the past week about adopting [github.com/actions](https://github.com/actions) in other contexts too, and we are indeed going to start adopting them more. We could go either way here. If you're OK being our first adopter of [actions/stale](https://github.com/actions/stale), I'll revise this PR to adopt it instead.

jeffhandley commented:
I prepared a branch that uses actions/stale and I did some testing in my [jeffhandley/github-action-testing](https://github.com/jeffhandley/github-action-testing) repository. I found a reason to hesitate about adopting `actions/stale` for this repo and other high-volume repositories. **With a high volume of issues and PRs, the action will encounter API rate limiting.** The action is built to save its state and resume from where it left off though.

The stale action is purely based on cron schedules and not event triggers. This means that in order for an issue to have the 'stale' label _removed_, the job has to walk through the list of issues/PRs looking for activity/comments. When tested on my [jeffhandley/github-action-testing](https://github.com/jeffhandley/github-action-testing/actions/runs/8807724222/job/24175419565) repo, which has 2 total issues and PRs, the action was using as many as 5 API operations per issue in each run, and it was logging the API rate limit counter as part of the output:

```
Statistics:
Processed issues: 2
New stale issues: 2
Added issues labels: 2
Added issues comments: 2
Fetched items: 2
Fetched items events: 2
Fetched items comments: 2
Operations performed: 10
Github API rate used: 10
Github API rate remaining: 4985; reset at: Tue Apr 23 2024 22:45:52 GMT+0000 (Coordinated Universal Time)
```

@marcpopMSFT / @MiYanni -- I found that there are some existing dotnet org repos using actions/stale, but they are much lower volume repositories. If you're up for it, we could try adopting actions/stale here and monitor it for a while to observe the output and see if this will work as we desire or if this will lead to rate limiting issues with this or other workflows.

For reference, here's what the actions/stale adoption would look like:

https://github.com/jeffhandley/sdk/blob/a14856217dc1235de33f7828a6dc5a7f27b8666f/.github/workflows/stale.yml#L1-L33

/cc @danmoseley

marcpopMSFT commented:
@jeffhandley am I reading that correctly as 5k api calls per day? Would that be 5 api calls per PR/issue that had an update since the last time it was run so as long as we had <1k edited issues per day, we'd be ok?

jeffhandley commented:
A new rate limit is granted _every hour_.  But the action requires [iterating over _all_ open issues and PRs](https://github.com/actions/stale/blob/main/src/classes/issues-processor.ts#L564-L585) each invocation, with multiple operations needed per issue/PR. With 2500 issues currently open in dotnet/sdk, it'll take several invocations to make it through the list. When the limit is hit, the action gracefully shuts down and keeps its place though, and each invocation picks up where it left off. When it finishes, it starts over.

The actions/stale documentation has some good notes about this too (that I just read): https://github.com/actions/stale/tree/main?tab=readme-ov-file#operations-per-run

> GitHub has a [rate limit](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting) and if reached will block these API calls for one hour (or API calls from other actions using the same user (a.k.a.: the github-token from the [repo-token](https://github.com/actions/stale/tree/main?tab=readme-ov-file#repo-token) option)).

We'd be at the mercy of what ever our rate limit is for our dotnet org plan. But the action gives us a couple of config settings we could tune as needed:

- [`operations-per-run`](https://github.com/actions/stale/tree/main?tab=readme-ov-file#operations-per-run) - Allowing us to limit the number of operations this action consumes from our rate limit (so that other actions don't fail)
- [`ascending`](https://github.com/actions/stale/tree/main?tab=readme-ov-file#ascending) - Allowing us to operate over issues/PRs oldest first (more likely to be stale).

I'm supportive of giving this a trial run here. We don't have other actions running in a continuous way on this repo, so we could even open the `operations-per-run` up to a pretty high number to let it make the first major pass through marking issues as stale.

https://github.com/dotnet/sdk/compare/main...jeffhandley:jeffhandley/stale-action?expand=1

marcpopMSFT commented:
If you want to do the trial run of actions/stale, we can try it on this repo first. It sounds like it just runs regularly until it's done so even though we are rate limited by hour, it should be able to finish the whole list each day. It doesn't even need to run that often but I think you have it configured to run hourly partly because it's going to be rate limited and have to pick back up the next hour.

Your linked change looks right to me (having no prior experience with actions/stale). I assume it automatically handles removing the label if there's a comment (that's what it says it does but your yml doesn't specify any of those issues)
